src/ucp/api/ucp_def.h

```diff
@@ -629,43 +629,49 @@ typedef ucs_status_t (*ucp_am_callback_t)(void *arg, void *data, size_t length,
  *
- * @return UCS_OK         @a data will not persist after the callback returns.
- *                        If UCP_AM_RECV_ATTR_FLAG_RNDV flag is set in
- *                        @a param->recv_attr and @ref ucp_am_recv_data_nbx was not
- *                        called for this data, the data descriptor will be
- *                        dropped and the corresponding @ref ucp_am_send_nbx
- *                        call should complete with UCS_OK status.
- *                        If:
- *                        1. UCP_AM_RECV_ATTR_FLAG_FIRST flag is set in
- *                        @a param->recv_attr
- *                        2. This receive handler was registered with
- *                        @a UCP_AM_FLAG_FIRST_MSG flag and
- *                        3. @ref ucp_am_recv_data_nbx was not called for this data,
- *                        the whole message will be dropped as well as this data
- *                        descriptor.
- *
- * @return UCS_INPROGRESS Can only be returned if @a param->recv_attr flags
- *                        contains UCP_AM_RECV_ATTR_FLAG_DATA or
- *                        UCP_AM_RECV_ATTR_FLAG_RNDV. The @a data will persist
- *                        after the callback has returned. To free the memory,
- *                        a pointer to the data must be passed into
- *                        @ref ucp_am_data_release or data receive is initiated by
- *                        @ref ucp_am_recv_data_nbx.
- *
- * @return otherwise      Can only be returned if @a param->recv_attr contains
- *                        UCP_AM_RECV_ATTR_FLAG_RNDV. In this case data
- *                        descriptor @a data will be dropped and the
- *                        corresponding @ref ucp_am_send_nbx call should
- *                        complete with the status returned from the callback.
+ * @return UCS_OK             @a data will not persist after the callback returns.
+ *                            If UCP_AM_RECV_ATTR_FLAG_RNDV flag is set in
+ *                            @a param->recv_attr and @ref ucp_am_recv_data_nbx was not
+ *                            called for this data, the data descriptor will be
+ *                            dropped and the corresponding @ref ucp_am_send_nbx
+ *                            call should complete with UCS_OK status.
+ *
+ * @return UCS_INPROGRESS     Can only be returned if @a param->recv_attr flags
+ *                            contains UCP_AM_RECV_ATTR_FLAG_DATA or
+ *                            UCP_AM_RECV_ATTR_FLAG_RNDV. The @a data will persist
+ *                            after the callback has returned. To free the memory,
+ *                            a pointer to the data must be passed into
+ *                            @ref ucp_am_data_release or data receive is initiated by
+ *                            @ref ucp_am_recv_data_nbx.
+ *
+ * @return UCS_ERR_NO_MESSAGE Can only be returned if this receive handler was
```

Comment:
maybe we don't have to support such a mode, at least not as the first stage. we can say that the application will be responsible to queue incoming fragments (potentially by using the context). WDYT?

Reply:
while it adds additional complexity to the API, I think it is quite useful feature for those users who needs message ordering, but does not really want to deal with fragments.